sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,368 @@
+# Bindless Images
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) Codeplay. All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Dependencies
+
+This extension is written against the Level Zero v1.5.8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+== API
+
+* Enumerations
++
+```
+ze_bindless_images_ext_version_t
+```
+* Structures
++
+```
+ze_image_mem_exp_desc_t
+ze_bindless_image_exp_desc_t
+ze_image_mem_alloc_exp_desc_t
+```
+
+== Overview
+
+Bindless images are a feature that provides flexibility on how images are accessed and used, such as removing limitations on how many images can be accessed as well as potentially improving performance.
```

Comment:
can you describe what are bindless images and how they differ from bindful images?

what is the goal of this extension ?

why current L0 API cannot be used?

If the main reason to have this extension is to be able to access image memory on HOST  with a pointer , or allocate an image with provided memory ( imported)  - then i do not understand benefits over RAW buffer memory -  without image format information, we cannot use tiled images or even linear images and samplers - images have their restrictions on row pitch / slice pitch  and dimensions.

we cannot import any memory and create image out of it .

if HOST  doe not need ot access memory - Image view  extension can redescribe image and create new image from existing / allocated memory. Wouldnt that be enough ?

Reply:
@HoppeMateusz Mateusz this is the main point - bindless textures are not bindless or bindful images (or buffers). It seems the word "bindless" has two different meaning and therefore people can get confused.
1. Bindless texture - L0 extension draft (here) - texture (image) not attached to the backing storage (thus not bound - bindless).
2. Bindless and bindful addressing model - two flavors of stateful addressing.

So this extension is very similar to the OpenCL 1.2 image from buffer extension, but with dynamic selection from which buffer image is created. And yes, it looks like only flat buffer memory could be supported, so no tiled memory layout as sampler capable platform could read it.

Reply:
event with flat linear memory - smapler may not be able to read it if pitch dimensions are not aligned with HW requirements - and from the intents in this spec - i see that  any mmoery can be used to create an image  - imported memory from different API - it means memory may not be big enough to have necessary padding or memory filled on HOST may not know excatct HW restrictions on row pith and slice pitch and paddings

Reply:
user defined pitch may not align with HW restrictions - if it doesn;t memory will have to be reallocated ( copied) - this is what we do in OpenCL when createImage with USE HOST PTR flag is used.

Restrictions depend on HW image format used - for RGBA restrictions may be different than for planar YUV image, or YCrCb images.

and once requirements are queried by user - the memory cannot be used ( in genreal) for image of tdifferent format - so user will have to allocate new memory - or stop using/ interpreting this memory  as described by first image formt..

Does spec cover all of this?

Reply:
In the case where the user has allocated with our extension to `zeMemAllocDevice` (though we are probably going to need to propose a new allocation function), the optimised pitch variables are returned to the user. This way the pitch will align with the HW requirements when creating the image. 

However, in the case where the user doesn't allocate with our extended allocation function, then it should be the responsibility of the user to query the hardware for the alignment requirements and make sure that their allocation is consistent with those requirements before creating the image.

What are the restrictions based on if not the number of bytes? The user should not be able to create an image on top of memory whose specified formats disagree with one another. We were also recently made aware of the Image View extension, what happens in the case where a user chooses to make an image view with a different image format than what the original image is defined with?

The spec doesn't cover these cases in detail enough. It's clear we're missing a lot and we are hoping to take your comments into consideration. Your feedback is greatly appreciated!

Reply:
from the code - looks that we do not check any compatibility - so any image view could be applied to existing image and L0 would not fail.